### PR TITLE
feat(fpl): add generalized success token FPL with parameterized base percentage

### PR DESCRIPTION
**Pull request by pemulis**
**Motivation**

We would like to have a generalized `SuccessToken` financial product library where the base percentage can be set as a parameter, with the remaining percentage acting like an embedded call option.

**Summary**

Using the `SimpleSuccessToken` as a model, this PR adds a generalized `SuccessToken` financial product library and associated tests.

**Details**

* We discussed creating an even more generalized financial product library, where the "option" part of the payout could follow arbitrary logic defined in an external financial product library, which could be a call option, binary option, range token, or something else. It was determined that this was actually harder to reason about as a deployer, in addition to being more difficult to test, so we stuck with the call option logic for the option part of the payout. The ultra-generalized version of this will be brought up in a Discourse thread for further discussion with the community.
* It might make sense to delete the `SimpleSuccessToken` contract and tests from the repo as part of this pull request, since the generalized version makes it redundant. We didn't do so yet, since that contract is still used for a proof-of-concept as part of an active deal pursuit. If we don't use the `SimpleSuccessToken` for a real contract, we should delete it later.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [x]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

**Issue(s)**

Fixes #XXXX

### CODE REVIEW

File: packages/core/contracts/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.sol [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+     */
+    function setLongShortPairParameters(
+        address longShortPair,
+        uint256 strikePrice,
+        uint256 basePercentage
+    ) public nonReentrant() {
+        require(ExpiringContractInterface(longShortPair).expirationTimestamp() != 0, "Invalid LSP address");
+        SuccessTokenLongShortPairParameters memory params = longShortPairParameters[longShortPair];
+        require(params.strikePrice == 0 && params.basePercentage == 0, "Parameters already set");
+        require(basePercentage != 0, "Base percentage cannot be set to 0");
```
Comment (nicholaspai) [inline packages/core/contracts/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.sol:44]: Should you also require `strikePrice != 0`? This way the previous `require` always reverts after the params are set. Also in your comments you say that the strike price cannot be 0

File: packages/core/contracts/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.sol [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+    mapping(address => SuccessTokenLongShortPairParameters) public longShortPairParameters;
+
+    /**
+     * @notice Enables any address to set the strike price for an associated LSP.
+     * @param longShortPair address of the LSP.
+     * @param strikePrice the strike price for the covered call for the associated LSP.
+     * @param basePercentage the base percentage of collateral per pair paid out to long tokens, expressed
+     * with 1e18 decimals. E.g., a 50% base percentage should be expressed 500000000000000000, or 0.5 with
+     * 1e18 decimals. The base percentage can not be set to 0.
```
Comment (nicholaspai) [inline packages/core/contracts/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.sol:30]:
```suggestion
     * 1e18 decimals. The base percentage cannot be set to 0.
```

File: packages/core/contracts/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.sol (hunk 1 above)
Comment (nicholaspai) [reply packages/core/contracts/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.sol:44]: Basically, repeat the logic of the `require` statement on line 66

File: packages/core/contracts/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.sol [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        uint256 positiveExpiryPrice = expiryPrice > 0 ? uint256(expiryPrice) : 0;
+        if (positiveExpiryPrice == 0 || positiveExpiryPrice <= params.strikePrice) return params.basePercentage;
+
+        // Else, token expires to be worth basePercentage + (1 - basePercentage) * (expiryPrice - strikePrice).
+        // E.g., if $TOKEN is $30 and strike is $20, long token is redeemable for 0.5 + 0.5*(30-20)/30 = 0.6667%
+        // which if the collateralPerPair is 2, is worth 1.3333 $TOKEN, which is worth $40 if 1 $TOKEN is worth $30.
+        // This return value is strictly < 1. The return value tends to 1 as the expiryPrice tends to infinity.
+        return
+            (
+                FixedPoint.Unsigned(params.basePercentage).add(
```
Comment (nicholaspai) [inline packages/core/contracts/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.sol:80]: Due to rounding down/precision errors this might return a slightly smaller value than expected. Are we ok with that?

File: packages/core/test/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.js [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+  beforeEach(async () => {
+    successTokenLSPFPL = await SuccessTokenLongShortPairFinancialProductLibrary.new().send({ from: accounts[0] });
+    lspMock = await LongShortPairMock.new(
+      "1000000", // _expirationTimestamp
+      collateralPerPair // _collateralPerPair
+    ).send({ from: accounts[0] });
+  });
+  describe("Long Short Pair Parameterization", () => {
+    it("Can set and fetch valid strikes", async () => {
```
Comment (nicholaspai) [inline packages/core/test/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.js:38]: Should add asserts that you can't set `strikePrice`/`basePercentage` equal to 0. General rule of thumb is that every single require statement should be tested once in a unit test.

File: packages/core/test/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.js [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        await didContractThrow(
+          successTokenLSPFPL.methods
+            .setLongShortPairParameters(ZERO_ADDRESS, strikePrice, basePercentage1)
+            .send({ from: accounts[0] })
+        )
+      );
+    });
+  });
+  describe("Compute expiry tokens for collateral with base percentage of 0.5", () => {
+    beforeEach(async () => {
```
Comment (nicholaspai) [inline packages/core/test/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.js:72]: Can you add a test that `percentageLongCollateralAtExpiry` reverts if the params are not set yet?

File: packages/core/test/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.js [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+      assert(
+        await didContractThrow(
+          successTokenLSPFPL.methods
+            .setLongShortPairParameters(ZERO_ADDRESS, strikePrice, basePercentage1)
+            .send({ from: accounts[0] })
+        )
+      );
+    });
+  });
+  describe("Compute expiry tokens for collateral with base percentage of 0.5", () => {
```
Comment (nicholaspai) [inline packages/core/test/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.js:71]: Can you add a test for when the expiry price = 0 since this is an edge case in the contract?

File: packages/core/test/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.js [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    });
+    it("Should never return a value greater than 1", async () => {
+      // Create a massive expiry price. 1e18*1e18. Under all conditions should return less than 1.
+      const expiryTokensForCollateral = await successTokenLSPFPL.methods
+        .percentageLongCollateralAtExpiry(toWei(toWei("1")))
+        .call({ from: lspMock.options.address });
+      assert.isTrue(toBN(expiryTokensForCollateral).lt(toBN(toWei("1"))));
+    });
+  });
+  describe("Compute expiry tokens for collateral with base percentage of 0.2", () => {
```
Comment (nicholaspai) [inline packages/core/test/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.js:106]: Can you add a test for when the expiry price = 0 since this is an edge case in the contract?

**Review by nicholaspai — COMMENTED** [review]
Thanks for this! I have one main contract comment and then requests to add unit tests to bolster code coverage.

File: packages/core/contracts/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.sol (hunk 3 above)
Comment (pemulis) [reply packages/core/contracts/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.sol:80]: My gut feeling is yes, but would be interested if @smb2796 disagrees. I'll add a note about this.

File: packages/core/contracts/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.sol [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        require(params.strikePrice != 0 || params.basePercentage != 0, "Params not set for calling LSP");
+
+        // If the expiry price is less than the strike price then the long options expire worthless (out of the money).
+        // In this case, return of value of the base percentage to the long tokenholders.
+        // Note we do not consider negative expiry prices in this implementation.
+        uint256 positiveExpiryPrice = expiryPrice > 0 ? uint256(expiryPrice) : 0;
+        if (positiveExpiryPrice == 0 || positiveExpiryPrice <= params.strikePrice) return params.basePercentage;
+
+        // Else, token expires to be worth basePercentage + (1 - basePercentage) * (expiryPrice - strikePrice).
+        // E.g., if $TOKEN is $30 and strike is $20, long token is redeemable for 0.5 + 0.5*(30-20)/30 = 0.6667%
```
Comment (chrismaree) [inline packages/core/contracts/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.sol:75]: the example you have here has `basePercentage` set to 0.5 (50%). perhaps add this to the comment?

File: packages/core/contracts/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.sol [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+import "./LongShortPairFinancialProductLibrary.sol";
+import "../../../../common/implementation/Lockable.sol";
+
+/**
+ * @title Success Token Long Short Pair Financial Product Library.
+ * @notice Adds settlement logic to create success token LSPs. A success token pays out a fixed amount of
+ * collateral as a floor, with the remaining amount functioning like an embedded option. The embedded
+ * option can use the payout logic of other financial product libraries, and resemble a covered call, a
+ * KPI option, a range bond, a capped yield dollar, or custom payout logic in a new financial library.
```
Comment (chrismaree) [inline packages/core/contracts/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.sol:12]: prehaps add the equation that uses the `strikePrice` `basepercentage` and `expiryPrice` to show how the `percentageLongCOllateral` is calculated in the comment here?

**Review by nicholaspai — APPROVED** [review]
All my comments are resolved, thanks! LGTM after @chrismaree 's comments are reviewed and after @smb2796 double checks one comment. I'll tag Sean.

**Review by chrismaree — APPROVED** [review]
This is great! thanks for updating the comments and all the nits. Really excited to see these used in the wild!

File: packages/core/contracts/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.sol [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+     */
+    function setLongShortPairParameters(
+        address longShortPair,
+        uint256 strikePrice,
+        uint256 basePercentage
+    ) public nonReentrant() {
+        require(ExpiringContractInterface(longShortPair).expirationTimestamp() != 0, "Invalid LSP address");
+        SuccessTokenLongShortPairParameters memory params = longShortPairParameters[longShortPair];
+        require(params.strikePrice == 0 && params.basePercentage == 0, "Parameters already set");
+        require(strikePrice != 0 && basePercentage != 0, "Base percentage and strike price cannot be set to 0");
```
Comment (chrismaree) [inline packages/core/contracts/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.sol:44]: One other question: this require will let you set `strikePrice` or `basePercentage` to 0 (assuming the other is not zero). is there anything bad that can happen in this case?

If `strikePrice` is zero then the SuccessToken is always in the money. if `basePercentage` is zero this acts just like an option. Neither of these seem "bad" but just wanted to see if there is anything else to think about in this parameterization context.

File: packages/core/contracts/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.sol (hunk 3 above)
Comment (smb2796) [reply packages/core/contracts/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.sol:82]: Seems fine to me. Is there even really a solution for this? Also this would be present in the covered call library as well, correct?

File: packages/core/contracts/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.sol (hunk 10 above)
Comment (pemulis) [reply packages/core/contracts/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.sol:44]: Hmm, I'm surprised that this would allow one parameter to be set to zero, since I thought the `&&` requires both to be non-zero?

File: packages/core/contracts/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.sol (hunk 3 above)
Comment (pemulis) [reply packages/core/contracts/financial-templates/common/financial-product-libraries/long-short-pair-libraries/SuccessTokenLongShortPairFinancialProductLibrary.sol:82]: The same issue would be there in covered call, too.
